### PR TITLE
FIX: compiling on vxworks

### PR DESCRIPTION
**Pull request by embedthis**
Changes to get the software to compile on VxWorks.

* VxWorks does not have stdint.h
* Different casting requirements for setsockopt and write

### CODE REVIEW

File: library/net.c [hunk 1]
```diff
@@ -203,7 +207,11 @@ int mbedtls_net_bind( mbedtls_net_context *ctx, const char *bind_ip, const char
 
         n = 1;
         if( setsockopt( ctx->fd, SOL_SOCKET, SO_REUSEADDR,
+#if VXWORKS
+                        (char *) &n, sizeof( n ) ) != 0 )
+#else
```
Comment (NWilson) [inline library/net.c]: No need for ifdefs on any of these three casts in this file: the cast to a (non-const) char* should be legal on all platforms, even though it's not required with modern const-correct headers.

**Comment by NWilson** [comment]
There are other platforms that don't have stdint.h either, such as Visual Studio 2005 (last version of VS with good support for Windows 2000).

I personally wouldn't add ifdefs around each #include of stdint.h, that's really pretty ugly. What we do in our builds is provide our own stdint.h on platforms that don't ship with it. I'd suggest doing that on VxWorks, rather than changing the mbedTLS code to accommodate old platforms.

**Comment by miklis** [comment]
Automatic CI verification build not done, please verify manually.

**Comment by mpg** [comment]
We announced in the 2.0.0 ChangeLog that we now require some minimal C99 support, including `long long` and fixed-size types such as `uint32_t`: https://github.com/ARMmbed/mbedtls/blob/d97f899f99e15fa058ade6d6d8ecf2bbcce74aee/ChangeLog#L283

I'm also inclined to think that it's the responsibility of the user, not us, to provide this header if their favourite toolchain doesn't.

File: library/net.c (hunk 1 above)
Comment (embedthis) [inline library/net.c]: Unfortunately not true in practice with the current shipping VxWorks 6.7 compiler.

I know, pretty ugly. The only other solution would be to introduce a typedef for the setsockopt arg. Equally ugly.

**Comment by embedthis** [comment]
Regarding C99 support. We could add those types in a VxWorks specific header.  The user is not in a place to fix the tool chain of the standard VxWorks compiler. So the we must do it if we want to support VxWorks. Otherwise, the user has to find, understand and fix the source in any VxWorks build.

**Comment by embedthis** [comment]
Where and how do you supply a VxWorks (stub) stdint.h?  Cannot put it under include/mbedtls, so where does it go?

File: library/net.c (hunk 1 above)
Comment (NWilson) [inline library/net.c]: I'm saying that the `(char*)` cast should be accepted by nonmal compilers (I think?), so we shouldn't need an ifdef, the `#if VXWORKS` lines here could well work on all platforms (it's worth a try).

**Comment by NWilson** [comment]
You can put something like this anywhere in your system (eg `~/include/stdint.h`), then `export CFLAGS=-I$HOME/include` when you configure mbedTLS. This is the stdint.h I use:

    /* simple stdint.h */
    #if defined(__sun) || defined(__hpux)
    #include <inttypes.h>

    #else
    /* Other UNIX platforms: define the types sensibly, trying to match the
     * definitions you'd find in the platform headers in later versions. */
    #if defined(_CHAR_IS_SIGNED)
    typedef char               int8_t;
    #else
    typedef signed char        int8_t;
    #endif
    typedef short              int16_t;
    typedef int                int32_t;
    typedef unsigned char      uint8_t;
    typedef unsigned short     uint16_t;
    typedef unsigned int       uint32_t;
    #if defined(_LP64) || defined(__LP64__)
    typedef long               int64_t;
    typedef unsigned long      uint64_t;
    #else
    typedef long long          int64_t;
    typedef unsigned long long uint64_t;
    #endif

    #endif

File: library/net.c (hunk 1 above)
Comment (embedthis) [inline library/net.c]: Yes, worth at test.

**Comment by embedthis** [comment]
So if this is not part of the standard Mbedtls software distribution, then the user cannot just build it using the standard build instructions.

I would propose that there be a standard os.h header that can include these kinds of things. That header can then be included by all source files. The os.h header can include stdint.h and other repeated includes.

File: library/net.c [hunk 2]
```diff
@@ -358,8 +366,13 @@ int mbedtls_net_accept( mbedtls_net_context *bind_ctx,
                          (struct sockaddr *) &local_addr, &n ) != 0 ||
```
Comment (venkatarajasekhar) [inline library/net.c]: Modified Code 
bind_ctx->fd = (int) socket( local_addr.ss_family,
                                            SOCK_DGRAM, IPPROTO_UDP );
bind_ctx->fd < 0 ||

**Comment by sbutcher-arm** [comment]
We've had this PR open a long time and @mpg and I have discussed it and agreed we would stay with the minimal requirement for C99 and stdint.h rather than try and accommodate all toolchains. If we adopt this PR, we'll also start adopting many other workarounds for other platforms complicating our testing and support, and as @NWilson and others have suggested, there are known workarounds which can be adopted.

**Pull request closed without merging**